modules.d/95virtiofs/mount-virtiofs.sh

```diff
@@ -0,0 +1,40 @@
+#!/usr/bin/sh
+
+virtiofs_fstype_opt() {
+	[ -n "$fstype" -a "$fstype" = "virtiofs" ]
+}
+
+virtiofs_root_opt() {
+	[ -n "$root" -a -z "${root%%virtiofs:*}" ]
+}
+
+virtiofs_root() {
+	[ virtiofs_fstype_opt -o virtiofs_root_opt ]
+}
+
+load_module() {
+	if ! (modprobe virtiofs || strstr "$(cat /proc/filesystems)" virtiofs); then
+		die "virtiofs is required but not available."
+	fi
+}
+
+mount_root() {
+	local fstag=${root#virtiofs:}
+	mount -t virtiofs -o "$rflags" "$fstag" "$NEWROOT" 2>&1 | vinfo
+}
+
+if virtiofs_root; then
+	load_module
```

Comment:
Btw, are all these small functions, each only called from a single place, really necessary? The code would be half of the current length without them and IMHO still perfectly understandable...